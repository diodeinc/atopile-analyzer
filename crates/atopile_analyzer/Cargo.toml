[package]
name = "atopile_analyzer"
version = "0.1.0"
edition = "2021"

[dependencies]
anyhow = "1.0.93"
assert_fs = "1.1.2"
atopile_parser = { path = "../atopile_parser" }
insta = { version = "1.41.1", features = ["filters", "yaml"] }
log = "0.4.22"
regex = "1.11.1"
serde = { version = "1.0.214", features = ["derive", "rc"] }
