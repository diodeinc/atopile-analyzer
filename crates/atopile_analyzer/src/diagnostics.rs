use std::{
    cell::{Ref, RefCell},
    collections::HashMap,
    path::{Path, PathBuf},
};

use crate::{evaluator::EvaluatorError, Location};

/// A diagnostic from the analyzer.
#[derive(Debug, Clone)]
pub struct AnalyzerDiagnostic {
    pub severity: AnalyzerDiagnosticSeverity,
    pub kind: AnalyzerDiagnosticKind,
    pub file: PathBuf,
}

#[derive(Debug, Copy, Clone)]
pub enum AnalyzerDiagnosticSeverity {
    Error,
    Warning,
}

#[derive(Debug, Clone)]
pub enum AnalyzerDiagnosticKind {
    UnconnectedInterface(UnconnectedInterfaceDiagnostic),
    Evaluator(EvaluatorError),
}

impl From<EvaluatorError> for AnalyzerDiagnostic {
    fn from(error: EvaluatorError) -> Self {
        let file = error.location.file.clone();
        Self {
            severity: AnalyzerDiagnosticSeverity::Error,
            kind: AnalyzerDiagnosticKind::Evaluator(error),
            file,
        }
    }
}
pub struct AnalyzerReporter {
    diagnostics: RefCell<HashMap<PathBuf, Vec<AnalyzerDiagnostic>>>,
}

impl AnalyzerReporter {
    pub fn new() -> Self {
        Self {
            diagnostics: RefCell::new(HashMap::new()),
        }
    }
}

impl Default for AnalyzerReporter {
    fn default() -> Self {
        Self::new()
    }
}

impl AnalyzerReporter {
    pub fn reset(&self) {
        self.diagnostics.borrow_mut().clear();
    }

    pub fn clear(&self, path: &Path) {
        self.diagnostics
            .borrow_mut()
            .entry(path.to_path_buf())
            .or_default()
            .clear();
    }

    pub fn report(&self, diagnostic: AnalyzerDiagnostic) {
        self.diagnostics
            .borrow_mut()
            .entry(diagnostic.file.clone())
            .or_default()
            .push(diagnostic);
    }

    pub fn diagnostics(&self) -> Ref<'_, HashMap<PathBuf, Vec<AnalyzerDiagnostic>>> {
        self.diagnostics.borrow()
    }
}

#[derive(Debug, Clone)]
pub struct UnconnectedInterfaceDiagnostic {
    pub instance_name: String,
    pub interface_name: String,

    pub instantiation_location: Location,
    pub interface_location: Location,
}
