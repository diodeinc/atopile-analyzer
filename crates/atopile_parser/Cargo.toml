[package]
name = "atopile_parser"
version = "0.1.0"
edition = "2021"

[dependencies]
chumsky = "0.9.3"
insta = "1.41.1"
serde = { version = "1.0.214", features = ["derive", "rc"] }

[dev-dependencies]
insta = "1.34.0"
glob = "0.3"
