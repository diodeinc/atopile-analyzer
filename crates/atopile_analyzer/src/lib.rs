pub mod diagnostics;
pub mod evaluator;

use std::{
    collections::HashMap,
    fmt::{Debug, Display},
    ops::Deref,
    path::{Path, PathBuf},
    sync::{Arc, Mutex},
};

use anyhow::{Context, Result};
use atopile_parser::{
    parser::{BlockStmt, Connectable, Expr, PortRef, Stmt, Symbol},
    AtopileSource, Position, Span, Spanned,
};
use evaluator::{resolve_import_path, Evaluator};
use log::{debug, info, warn};
use serde::Serialize;

use diagnostics::*;

pub use crate::evaluator::EvaluatorState;

#[derive(Debug, Clone, Serialize)]
pub struct Location {
    // TODO: Windows and Unix paths don't play nice together in snapshot tests, so just skip
    // serialization for now.
    #[serde(skip)]
    pub file: PathBuf,
    pub range: Range,
}

impl Display for Location {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}:{}", self.file.display(), self.range.start)
    }
}

#[derive(Debug, Clone, Copy, Serialize)]
pub struct Range {
    pub start: Position,
    pub end: Position,
}

#[derive(Debug)]
pub struct Located<T>(T, Location);

impl<T> Deref for Located<T> {
    type Target = T;
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

trait AsLocation {
    fn as_location(&self) -> Location;
}

impl<T> AsLocation for Located<T> {
    fn as_location(&self) -> Location {
        self.1.clone()
    }
}

impl AsLocation for Location {
    fn as_location(&self) -> Location {
        self.clone()
    }
}

impl<T> Located<T> {
    pub fn new(item: T, location: Location) -> Self {
        Self(item, location)
    }

    pub fn from_spanned(spanned: Spanned<T>, source: &AtopileSource, path: &Path) -> Self {
        let location = Location {
            file: path.to_path_buf(),
            range: Range {
                start: source.index_to_position(spanned.span().start),
                end: source.index_to_position(spanned.span().end),
            },
        };

        Self::new(spanned.take(), location)
    }

    pub fn location(&self) -> &Location {
        &self.1
    }
}

impl<T> From<(T, Location)> for Located<T> {
    fn from((item, location): (T, Location)) -> Self {
        Self(item, location)
    }
}

impl<T: ToOwned> Located<&T> {
    pub fn to_owned(&self) -> Located<T::Owned> {
        Located(self.0.to_owned(), self.1.clone())
    }
}

trait IntoLocated<T> {
    fn into_located(self, source: &AtopileSource) -> Located<T>;
}

impl<T> IntoLocated<T> for Spanned<T> {
    fn into_located(self, source: &AtopileSource) -> Located<T> {
        Located::from_spanned(self, source, source.path())
    }
}

trait IntoLocation {
    fn to_location(&self, source: &AtopileSource) -> Location;
}

impl IntoLocation for Span {
    fn to_location(&self, source: &AtopileSource) -> Location {
        Location {
            file: source.path().to_path_buf(),
            range: Range {
                start: source.index_to_position(self.start),
                end: source.index_to_position(self.end),
            },
        }
    }
}

/// A result from a goto definition request.
#[derive(Debug)]
pub struct GotoDefinitionResult {
    pub file: PathBuf,
    pub source_range: Range,
    pub target_range: Range,
    pub target_selection_range: Range,
}

trait AtopileSourceExt {
    #[allow(clippy::only_used_in_recursion)]
    fn port_ref_at_in_expr<'a>(
        &self,
        index: usize,
        expr: &'a Spanned<Expr>,
    ) -> Option<&'a Spanned<PortRef>>;

    /// Returns a PortRef that is at the given index into the source file, if there is one.
    #[allow(dead_code)]
    fn port_ref_at(&self, index: usize) -> Option<&Spanned<PortRef>>;

    #[allow(clippy::only_used_in_recursion)]
    fn symbol_name_at_in_expr<'a>(
        &self,
        index: usize,
        expr: &'a Spanned<Expr>,
    ) -> Option<&'a Spanned<Symbol>>;

    /// Returns a Spanned<String> for a symbol name that is at the given index
    /// into the source file, if there is one.
    fn symbol_name_at(&self, index: usize) -> Option<&Spanned<Symbol>>;

    fn file_path_at(&self, index: usize) -> Option<&Spanned<String>>;
}

impl AtopileSourceExt for AtopileSource {
    #[allow(clippy::only_used_in_recursion)]
    fn port_ref_at_in_expr<'a>(
        &self,
        index: usize,
        expr: &'a Spanned<Expr>,
    ) -> Option<&'a Spanned<PortRef>> {
        match &expr.deref() {
            Expr::Port(port) => {
                if port.span().contains(&index) {
                    Some(port)
                } else {
                    None
                }
            }
            Expr::BinaryOp(binary_op) => {
                if binary_op.left.span().contains(&index) {
                    self.port_ref_at_in_expr(index, &binary_op.left)
                } else if binary_op.right.span().contains(&index) {
                    self.port_ref_at_in_expr(index, &binary_op.right)
                } else {
                    None
                }
            }
            _ => None,
        }
    }

    /// Returns a PortRef that is at the given index into the source file, if there is one.
    fn port_ref_at(&self, index: usize) -> Option<&Spanned<PortRef>> {
        let stmt = self.stmt_at(index)?;
        match &stmt.deref() {
            Stmt::Assign(assign) => {
                if assign.target.span().contains(&index) {
                    Some(&assign.target)
                } else {
                    self.port_ref_at_in_expr(index, &assign.value)
                }
            }
            Stmt::Specialize(specialize) => {
                if specialize.port.span().contains(&index) {
                    Some(&specialize.port)
                } else {
                    None
                }
            }
            Stmt::Connect(connect) => {
                if let Connectable::Port(port) = &connect.left.deref() {
                    if port.span().contains(&index) {
                        Some(port)
                    } else {
                        None
                    }
                } else if let Connectable::Port(port) = &connect.right.deref() {
                    if port.span().contains(&index) {
                        Some(port)
                    } else {
                        None
                    }
                } else {
                    None
                }
            }
            Stmt::Assert(assert) => self.port_ref_at_in_expr(index, &assert.expr),
            _ => None,
        }
    }

    #[allow(clippy::only_used_in_recursion)]
    fn symbol_name_at_in_expr<'a>(
        &self,
        index: usize,
        expr: &'a Spanned<Expr>,
    ) -> Option<&'a Spanned<Symbol>> {
        match &expr.deref() {
            Expr::New(symbol) => symbol.span().contains(&index).then_some(symbol),
            Expr::BinaryOp(binary_op) => {
                if binary_op.left.span().contains(&index) {
                    self.symbol_name_at_in_expr(index, &binary_op.left)
                } else if binary_op.right.span().contains(&index) {
                    self.symbol_name_at_in_expr(index, &binary_op.right)
                } else {
                    None
                }
            }
            _ => None,
        }
    }

    /// Returns a Spanned<String> for a symbol name that is at the given index
    /// into the source file, if there is one.
    fn symbol_name_at(&self, index: usize) -> Option<&Spanned<Symbol>> {
        let stmt = self.stmt_at(index)?;
        match &stmt.deref() {
            Stmt::Import(import) => import.imports.iter().find(|i| i.span().contains(&index)),
            Stmt::DepImport(import) => import.name.span().contains(&index).then_some(&import.name),
            Stmt::Assign(assign) => self.symbol_name_at_in_expr(index, &assign.value),
            Stmt::Specialize(specialize) => specialize
                .value
                .span()
                .contains(&index)
                .then_some(&specialize.value),
            Stmt::Block(block) => block
                .parent
                .as_ref()
                .and_then(|p| p.span().contains(&index).then_some(p)),
            _ => None,
        }
    }

    fn file_path_at(&self, index: usize) -> Option<&Spanned<String>> {
        let stmt = self.stmt_at(index)?;
        match &stmt.deref() {
            Stmt::Import(import) => import
                .from_path
                .span()
                .contains(&index)
                .then_some(&import.from_path),
            Stmt::DepImport(import) => import
                .from_path
                .span()
                .contains(&index)
                .then_some(&import.from_path),
            _ => None,
        }
    }
}
pub(crate) struct FileCache {
    files: Mutex<HashMap<PathBuf, FileCacheEntry>>,
}

#[derive(Clone)]
struct FileCacheEntry {
    source: Arc<AtopileSource>,
}

impl FileCacheEntry {
    pub fn new(source: Arc<AtopileSource>) -> Self {
        Self { source }
    }
}

impl FileCache {
    pub fn new() -> Self {
        Self {
            files: Mutex::new(HashMap::new()),
        }
    }

    pub fn get(&self, path: &Path) -> Option<Arc<AtopileSource>> {
        self.files
            .lock()
            .unwrap()
            .get(path)
            .map(|e| e.source.clone())
    }

    pub fn insert(&self, path: PathBuf, source: Arc<AtopileSource>) {
        self.files
            .lock()
            .unwrap()
            .insert(path, FileCacheEntry::new(source));
    }

    pub fn remove(&self, path: &Path) {
        self.files.lock().unwrap().remove(path);
    }
}

pub struct AtopileAnalyzer {
    files: Arc<FileCache>,
    evaluator: Evaluator,
    open_files: std::collections::HashSet<PathBuf>,
}

impl AtopileAnalyzer {
    pub fn new() -> Self {
        let files = Arc::new(FileCache::new());
        Self {
            files: files.clone(),
            evaluator: Evaluator::default(),
            open_files: std::collections::HashSet::new(),
        }
    }
}

impl Default for AtopileAnalyzer {
    fn default() -> Self {
        Self::new()
    }
}

impl AtopileAnalyzer {
    #[cfg(test)]
    pub fn evaluate_source_for_test(&self, source: &AtopileSource) -> EvaluatorState {
        let mut evaluator = Evaluator::default();
        evaluator.set_source(source.path(), Arc::new(source.clone()));
        evaluator.state().clone()
    }

    /// Load the source file at the given path. Will first check the in-memory cache maintained by
    /// `set_source`, and if not found, read from the filesystem.
    fn load_source(&self, path: &PathBuf) -> Result<Arc<AtopileSource>> {
        debug!("loading source: {:?}", path);
        let path = path.canonicalize()?;

        if let Some(source) = self.files.get(&path) {
            debug!("source already loaded: {:?}", path);
            return Ok(source.clone());
        }

        debug!("loading source from disk: {:?}", path);
        let content = std::fs::read_to_string(&path).context("Failed to read source file")?;
        let source = AtopileSource::new(content, path.to_path_buf());
        Ok(Arc::new(source))
    }

    /// Set the source file at the given path.
    pub fn set_source(&mut self, path: &Path, source: Arc<AtopileSource>) -> Result<()> {
        let path = path.canonicalize()?;
        self.files.insert(path.clone(), source.clone());
        self.evaluator.set_source(&path, source.clone());
        Ok(())
    }

    /// Remove the source file at the given path.
    pub fn remove_source(&mut self, path: &Path) -> Result<()> {
        self.files.remove(&path.canonicalize()?);
        self.open_files.remove(&path.canonicalize()?);
        self.evaluator.remove_source(path);
        Ok(())
    }

    /// Mark a file as open in the editor.
    pub fn mark_file_open(&mut self, path: &Path) -> Result<()> {
        self.open_files.insert(path.canonicalize()?);
        Ok(())
    }

    /// Mark a file as closed in the editor.
    pub fn mark_file_closed(&mut self, path: &Path) -> Result<()> {
        self.open_files.remove(&path.canonicalize()?);
        Ok(())
    }

    /// Get all open files.
    pub fn get_open_files(&self) -> &std::collections::HashSet<PathBuf> {
        &self.open_files
    }

    /// Run all diagnostics.
    pub fn diagnostics(&mut self) -> Result<Vec<AnalyzerDiagnostic>> {
        let mut diagnostics = vec![];

        diagnostics.extend(
            self.evaluator
                .reporter()
                .diagnostics()
                .values()
                .flat_map(|v| v.iter())
                .cloned(),
        );

        Ok(diagnostics)
    }

    /// Find the BlockStmt that defines the given name, traversing through imports as necessary.
    fn find_definition(
        &self,
        source: &AtopileSource,
        name: &str,
    ) -> Result<Option<Located<BlockStmt>>> {
        if let Some(block) = self.find_definition_in_source(source, name) {
            // The definition is in this file, so just return it.
            Ok(Some(Located::from_spanned(
                block.to_owned(),
                source,
                source.path(),
            )))
        } else {
            // Let's see if we import this symbol, then recurse.
            debug!("looking for import for {:?}", name);
            let imported_file = source
                .traverse_all_stmts()
                .filter_map(|(stmt, _)| match stmt.deref() {
                    Stmt::Import(import) => Some((
                        import.from_path.deref(),
                        import.imports.iter().map(|i| i.deref().clone()).collect(),
                    )),
                    Stmt::DepImport(import) => {
                        Some((import.from_path.deref(), vec![import.name.deref().clone()]))
                    }
                    _ => None,
                })
                .find(|(_import_path, imports)| imports.iter().any(|i| i.deref() == name))
                .map(|(import_path, _imports)| {
                    let path = resolve_import_path(source.path(), Path::new(import_path));
                    debug!("resolved import path: {:?}", path);
                    path
                })
                .context(format!("failed to resolve import path for {:?}", name))?
                .map(|import| self.load_source(&import))
                .transpose()
                .context(format!("failed to load source for import {:?}", name))?;

            if let Some(imported_file) = imported_file {
                debug!("found imported file: {:?}", imported_file.path());
                self.find_definition(&imported_file, name)
            } else {
                warn!(
                    "can't find definition for {:?}: no matching import found",
                    name
                );
                Ok(None)
            }
        }
    }

    /// Find where a block with the given name is defined in this source file.
    fn find_definition_in_source<'a>(
        &self,
        source: &'a AtopileSource,
        name: &str,
    ) -> Option<Spanned<&'a BlockStmt>> {
        source
            .ast()
            .iter()
            .filter_map(|s| match s.deref() {
                Stmt::Block(block) => Some((block, s.span().clone()).into()),
                _ => None,
            })
            .find(|b: &Spanned<&BlockStmt>| b.deref().name.deref().to_string() == name)
    }

    /// Create a GotoDefinitionResult for the path component of an import, e.g. `path/to/file.ato`
    /// here:
    /// ```ato
    /// from "path/to/file.ato" import Symbol
    /// ```
    fn handle_goto_definition_path(
        &self,
        source: &AtopileSource,
        source_path: &Path,
        path_token: &Spanned<String>,
    ) -> Result<Option<GotoDefinitionResult>> {
        let source_range_start = source.index_to_position(path_token.span().start);
        let source_range_end = source.index_to_position(path_token.span().end);

        let resolved_path = resolve_import_path(source_path, Path::new(path_token.deref()))
            .context(format!(
                "failed to resolve import path for {:?}",
                path_token
            ))?;

        Ok(Some(GotoDefinitionResult {
            file: resolved_path,
            source_range: Range {
                start: source_range_start,
                end: source_range_end,
            },
            target_range: Range {
                start: Position { line: 0, column: 0 },
                end: Position { line: 0, column: 0 },
            },
            target_selection_range: Range {
                start: Position { line: 0, column: 0 },
                end: Position { line: 0, column: 0 },
            },
        }))
    }

    fn handle_goto_definition_for_symbol(
        &self,
        source: &AtopileSource,
        symbol: &Spanned<Symbol>,
    ) -> Result<Option<GotoDefinitionResult>> {
        let def = self.find_definition(source, symbol.deref())?;

        if let Some(def) = def {
            Ok(Some(GotoDefinitionResult {
                file: def.location().file.clone(),
                source_range: Range {
                    start: source.index_to_position(symbol.span().start),
                    end: source.index_to_position(symbol.span().end),
                },
                target_range: def.location().range,
                target_selection_range: def.location().range,
            }))
        } else {
            Ok(None)
        }
    }

    /// Attempt to resolve a goto definition request at the given position.
    pub fn goto_definition(
        &self,
        path: &PathBuf,
        position: Position,
    ) -> Result<Option<GotoDefinitionResult>> {
        let source = self.load_source(path)?;

        let index = source.position_to_index(position);
        let symbol = source.symbol_name_at(index);
        let file_path = source.file_path_at(index);

        if let Some(symbol) = symbol {
            info!("goto definition for symbol: {:?}", symbol);
            self.handle_goto_definition_for_symbol(&source, symbol)
        } else if let Some(file_path) = file_path {
            info!("goto definition for path: {:?}", file_path);
            self.handle_goto_definition_path(&source, path, file_path)
        } else {
            info!("no goto definition found");
            Ok(None)
        }
    }

    pub fn get_netlist(&mut self) -> &EvaluatorState {
        self.evaluator.resolve_reference_designators();
        self.evaluator.state()
    }
}
