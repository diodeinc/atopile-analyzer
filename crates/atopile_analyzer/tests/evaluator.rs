use atopile_analyzer::diagnostics::{
    AnalyzerDiagnostic, AnalyzerDiagnosticKind, AnalyzerDiagnosticSeverity,
};
use atopile_analyzer::evaluator::Evaluator;
use atopile_parser::AtopileSource;
use serde::Serialize;
use std::fs;
use std::path::PathBuf;
use std::sync::Arc;

#[derive(Debug, Serialize)]
struct DiagnosticInfo {
    severity: String,
    kind: String,
    file: String,
}

impl From<&AnalyzerDiagnostic> for DiagnosticInfo {
    fn from(diag: &AnalyzerDiagnostic) -> Self {
        let severity = match diag.severity {
            AnalyzerDiagnosticSeverity::Error => "Error",
            AnalyzerDiagnosticSeverity::Warning => "Warning",
        };

        let kind = match &diag.kind {
            AnalyzerDiagnosticKind::UnconnectedInterface(_) => "UnconnectedInterface",
            AnalyzerDiagnosticKind::Evaluator(err) => &format!("Evaluator: {}", err),
        };

        Self {
            severity: severity.to_string(),
            kind: kind.to_string(),
            file: diag.file.to_string_lossy().to_string(),
        }
    }
}

#[derive(Debug, Serialize)]
struct EvaluatorTestResult {
    state: atopile_analyzer::evaluator::EvaluatorState,
    diagnostics: Vec<DiagnosticInfo>,
}

macro_rules! create_evaluator_test {
    ($name:ident) => {
        #[test]
        fn $name() {
            let input = fs::read_to_string(concat!(
                "tests/resources/corpus/",
                stringify!($name),
                ".ato"
            ))
            .unwrap();
            let normalized_input = input.replace("\r\n", "\n");

            let file_path = concat!("tests/resources/corpus/", stringify!($name), ".ato");
            let path_buf = PathBuf::from(file_path);

            let source = Arc::new(AtopileSource::new(
                normalized_input.to_string(),
                path_buf.clone(),
            ));

            let mut evaluator = Evaluator::default();
            evaluator.set_source(&path_buf, source);
            let state = evaluator.evaluate();

            let diagnostics = evaluator.reporter()
                .diagnostics()
                .get(&path_buf)
                .map_or_else(Vec::new, |diags| {
                    diags.iter().map(DiagnosticInfo::from).collect()
                });

            let result = EvaluatorTestResult {
                state,
                diagnostics,
            };

            insta::with_settings!({
                sort_maps => true
            }, {
                insta::assert_yaml_snapshot!(result);
            });
        }
    };
}

create_evaluator_test!(simple_module);
create_evaluator_test!(simple_component);
create_evaluator_test!(simple_connection);
create_evaluator_test!(forward_reference);
create_evaluator_test!(cyclic_inheritance);
create_evaluator_test!(duplicate_declaration);
create_evaluator_test!(pin_connections);
create_evaluator_test!(dependency_ordering);
