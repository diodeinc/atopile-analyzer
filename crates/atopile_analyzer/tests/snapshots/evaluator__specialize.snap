---
source: crates/atopile_analyzer/tests/evaluator.rs
expression: result
---
state:
  instances:
    "tests/resources/corpus/specialize.ato:BaseInterface":
      type_ref:
        source_path: tests/resources/corpus/specialize.ato
        module_name: BaseInterface
      kind: Module
      attributes: {}
      children:
        a: "tests/resources/corpus/specialize.ato:BaseInterface.a"
        b: "tests/resources/corpus/specialize.ato:BaseInterface.b"
      connections: []
      reference_designator: ~
    "tests/resources/corpus/specialize.ato:BaseInterface.a":
      type_ref:
        source_path: ""
        module_name: ""
      kind: Port
      attributes: {}
      children: {}
      connections: []
      reference_designator: ~
    "tests/resources/corpus/specialize.ato:BaseInterface.b":
      type_ref:
        source_path: ""
        module_name: ""
      kind: Port
      attributes: {}
      children: {}
      connections: []
      reference_designator: ~
    "tests/resources/corpus/specialize.ato:ExtendedInterface":
      type_ref:
        source_path: tests/resources/corpus/specialize.ato
        module_name: ExtendedInterface
      kind: Module
      attributes: {}
      children:
        a: "tests/resources/corpus/specialize.ato:ExtendedInterface.a"
        b: "tests/resources/corpus/specialize.ato:ExtendedInterface.b"
        c: "tests/resources/corpus/specialize.ato:ExtendedInterface.c"
        d: "tests/resources/corpus/specialize.ato:ExtendedInterface.d"
      connections:
        - left: "tests/resources/corpus/specialize.ato:ExtendedInterface.a"
          right: "tests/resources/corpus/specialize.ato:ExtendedInterface.b"
      reference_designator: ~
    "tests/resources/corpus/specialize.ato:ExtendedInterface.a":
      type_ref:
        source_path: ""
        module_name: ""
      kind: Port
      attributes: {}
      children: {}
      connections: []
      reference_designator: ~
    "tests/resources/corpus/specialize.ato:ExtendedInterface.b":
      type_ref:
        source_path: ""
        module_name: ""
      kind: Port
      attributes: {}
      children: {}
      connections: []
      reference_designator: ~
    "tests/resources/corpus/specialize.ato:ExtendedInterface.c":
      type_ref:
        source_path: ""
        module_name: ""
      kind: Port
      attributes: {}
      children: {}
      connections: []
      reference_designator: ~
    "tests/resources/corpus/specialize.ato:ExtendedInterface.d":
      type_ref:
        source_path: ""
        module_name: ""
      kind: Port
      attributes: {}
      children: {}
      connections: []
      reference_designator: ~
    "tests/resources/corpus/specialize.ato:NestedModule":
      type_ref:
        source_path: tests/resources/corpus/specialize.ato
        module_name: NestedModule
      kind: Module
      attributes: {}
      children:
        iface: "tests/resources/corpus/specialize.ato:NestedModule.iface"
        sub: "tests/resources/corpus/specialize.ato:NestedModule.sub"
      connections:
        - left: "tests/resources/corpus/specialize.ato:NestedModule.iface.a"
          right: "tests/resources/corpus/specialize.ato:NestedModule.sub.a"
      reference_designator: ~
    "tests/resources/corpus/specialize.ato:NestedModule.iface":
      type_ref:
        source_path: tests/resources/corpus/specialize.ato
        module_name: BaseInterface
      kind: Module
      attributes: {}
      children:
        a: "tests/resources/corpus/specialize.ato:NestedModule.iface.a"
        b: "tests/resources/corpus/specialize.ato:NestedModule.iface.b"
      connections: []
      reference_designator: ~
    "tests/resources/corpus/specialize.ato:NestedModule.iface.a":
      type_ref:
        source_path: ""
        module_name: ""
      kind: Port
      attributes: {}
      children: {}
      connections: []
      reference_designator: ~
    "tests/resources/corpus/specialize.ato:NestedModule.iface.b":
      type_ref:
        source_path: ""
        module_name: ""
      kind: Port
      attributes: {}
      children: {}
      connections: []
      reference_designator: ~
    "tests/resources/corpus/specialize.ato:NestedModule.sub":
      type_ref:
        source_path: tests/resources/corpus/specialize.ato
        module_name: BaseInterface
      kind: Module
      attributes: {}
      children:
        a: "tests/resources/corpus/specialize.ato:NestedModule.sub.a"
        b: "tests/resources/corpus/specialize.ato:NestedModule.sub.b"
      connections: []
      reference_designator: ~
    "tests/resources/corpus/specialize.ato:NestedModule.sub.a":
      type_ref:
        source_path: ""
        module_name: ""
      kind: Port
      attributes: {}
      children: {}
      connections: []
      reference_designator: ~
    "tests/resources/corpus/specialize.ato:NestedModule.sub.b":
      type_ref:
        source_path: ""
        module_name: ""
      kind: Port
      attributes: {}
      children: {}
      connections: []
      reference_designator: ~
    "tests/resources/corpus/specialize.ato:TestModule":
      type_ref:
        source_path: tests/resources/corpus/specialize.ato
        module_name: TestModule
      kind: Module
      attributes: {}
      children:
        iface: "tests/resources/corpus/specialize.ato:TestModule.iface"
        nested: "tests/resources/corpus/specialize.ato:TestModule.nested"
      connections: []
      reference_designator: ~
    "tests/resources/corpus/specialize.ato:TestModule.iface":
      type_ref:
        source_path: tests/resources/corpus/specialize.ato
        module_name: ExtendedInterface
      kind: Module
      attributes: {}
      children:
        a: "tests/resources/corpus/specialize.ato:TestModule.iface.a"
        b: "tests/resources/corpus/specialize.ato:TestModule.iface.b"
        c: "tests/resources/corpus/specialize.ato:TestModule.iface.c"
        d: "tests/resources/corpus/specialize.ato:TestModule.iface.d"
      connections:
        - left: "tests/resources/corpus/specialize.ato:TestModule.iface.a"
          right: "tests/resources/corpus/specialize.ato:TestModule.iface.b"
      reference_designator: ~
    "tests/resources/corpus/specialize.ato:TestModule.iface.a":
      type_ref:
        source_path: ""
        module_name: ""
      kind: Port
      attributes: {}
      children: {}
      connections: []
      reference_designator: ~
    "tests/resources/corpus/specialize.ato:TestModule.iface.b":
      type_ref:
        source_path: ""
        module_name: ""
      kind: Port
      attributes: {}
      children: {}
      connections: []
      reference_designator: ~
    "tests/resources/corpus/specialize.ato:TestModule.iface.c":
      type_ref:
        source_path: ""
        module_name: ""
      kind: Port
      attributes: {}
      children: {}
      connections: []
      reference_designator: ~
    "tests/resources/corpus/specialize.ato:TestModule.iface.d":
      type_ref:
        source_path: ""
        module_name: ""
      kind: Port
      attributes: {}
      children: {}
      connections: []
      reference_designator: ~
    "tests/resources/corpus/specialize.ato:TestModule.nested":
      type_ref:
        source_path: tests/resources/corpus/specialize.ato
        module_name: NestedModule
      kind: Module
      attributes: {}
      children:
        iface: "tests/resources/corpus/specialize.ato:TestModule.nested.iface"
        sub: "tests/resources/corpus/specialize.ato:TestModule.nested.sub"
      connections:
        - left: "tests/resources/corpus/specialize.ato:TestModule.nested.iface.a"
          right: "tests/resources/corpus/specialize.ato:TestModule.nested.sub.a"
      reference_designator: ~
    "tests/resources/corpus/specialize.ato:TestModule.nested.iface":
      type_ref:
        source_path: tests/resources/corpus/specialize.ato
        module_name: ExtendedInterface
      kind: Module
      attributes: {}
      children:
        a: "tests/resources/corpus/specialize.ato:TestModule.nested.iface.a"
        b: "tests/resources/corpus/specialize.ato:TestModule.nested.iface.b"
        c: "tests/resources/corpus/specialize.ato:TestModule.nested.iface.c"
        d: "tests/resources/corpus/specialize.ato:TestModule.nested.iface.d"
      connections:
        - left: "tests/resources/corpus/specialize.ato:TestModule.nested.iface.a"
          right: "tests/resources/corpus/specialize.ato:TestModule.nested.iface.b"
      reference_designator: ~
    "tests/resources/corpus/specialize.ato:TestModule.nested.iface.a":
      type_ref:
        source_path: ""
        module_name: ""
      kind: Port
      attributes: {}
      children: {}
      connections: []
      reference_designator: ~
    "tests/resources/corpus/specialize.ato:TestModule.nested.iface.b":
      type_ref:
        source_path: ""
        module_name: ""
      kind: Port
      attributes: {}
      children: {}
      connections: []
      reference_designator: ~
    "tests/resources/corpus/specialize.ato:TestModule.nested.iface.c":
      type_ref:
        source_path: ""
        module_name: ""
      kind: Port
      attributes: {}
      children: {}
      connections: []
      reference_designator: ~
    "tests/resources/corpus/specialize.ato:TestModule.nested.iface.d":
      type_ref:
        source_path: ""
        module_name: ""
      kind: Port
      attributes: {}
      children: {}
      connections: []
      reference_designator: ~
    "tests/resources/corpus/specialize.ato:TestModule.nested.sub":
      type_ref:
        source_path: tests/resources/corpus/specialize.ato
        module_name: BaseInterface
      kind: Module
      attributes: {}
      children:
        a: "tests/resources/corpus/specialize.ato:TestModule.nested.sub.a"
        b: "tests/resources/corpus/specialize.ato:TestModule.nested.sub.b"
      connections: []
      reference_designator: ~
    "tests/resources/corpus/specialize.ato:TestModule.nested.sub.a":
      type_ref:
        source_path: ""
        module_name: ""
      kind: Port
      attributes: {}
      children: {}
      connections: []
      reference_designator: ~
    "tests/resources/corpus/specialize.ato:TestModule.nested.sub.b":
      type_ref:
        source_path: ""
        module_name: ""
      kind: Port
      attributes: {}
      children: {}
      connections: []
      reference_designator: ~
diagnostics: []
